### PR TITLE
adds create directory method

### PR DESCRIPTION
**Pull request by artemredkin**
Motivation:
Creating directories is a common task.

Modifications:
 - Adds mkdir syscall
 - Adds create directory implementation
 - Test

### CODE REVIEW

File: Sources/NIOPosix/NonBlockingFileIO.swift [hunk 1]
```diff
@@ -687,6 +687,66 @@ public struct NonBlockingFileIO: Sendable {
...
+
+        // Slow path
+        var index = path.index(before: path.endIndex)
+        while index != path.startIndex && path[index] == "/" {
+            index = path.index(before: index)
+        }
+
+        while index != path.startIndex, path[index] != "/" {
+            index = path.index(before: index)
+        }
```
Comment (Lukasa) [inline Sources/NIOPosix/NonBlockingFileIO.swift:711]: Can we move all of this computation onto the UTF8-view of the String?

File: Sources/NIOPosix/NonBlockingFileIO.swift [hunk 2]
```diff
@@ -687,6 +687,66 @@ public struct NonBlockingFileIO: Sendable {
...
+            if (S_IFMT & s.st_mode) == S_IFDIR {
+                return
+            }
+            throw IOError(errnoCode: ENOTDIR, reason: "Not a directory")
+        } catch {
+            // ignore error
+        }
+
+        // Slow path
+        var index = path.index(before: path.endIndex)
```
Comment (Lukasa) [inline Sources/NIOPosix/NonBlockingFileIO.swift:704]: You need an early-exit here if the path is empty.

File: Sources/NIOPosix/NonBlockingFileIO.swift (hunk 1 above)
Comment (Lukasa) [inline Sources/NIOPosix/NonBlockingFileIO.swift:711]: Can you also explain what these two blocks are doing?

File: Sources/NIOPosix/NonBlockingFileIO.swift [hunk 3]
```diff
@@ -687,6 +687,66 @@ public struct NonBlockingFileIO: Sendable {
...
+        while index != path.startIndex && path[index] == "/" {
+            index = path.index(before: index)
+        }
+
+        while index != path.startIndex, path[index] != "/" {
```
Comment (Lukasa) [inline Sources/NIOPosix/NonBlockingFileIO.swift:709]: Prefer `&&` to `,`

File: Sources/NIOPosix/NonBlockingFileIO.swift [hunk 4]
```diff
@@ -687,6 +687,70 @@ public struct NonBlockingFileIO: Sendable {
...
+
+        do {
+            try Posix.mkdir(pathname: path, mode: mode)
+        } catch {
+            do {
+                var s = stat()
+                try Posix.lstat(pathname: path, outStat: &s)
+                if (S_IFMT & s.st_mode) == S_IFDIR {
+                    return
+                }
```
Comment (Lukasa) [inline Sources/NIOPosix/NonBlockingFileIO.swift:725]: It'd be good to explain what this block is trying to do.

File: Sources/NIOPosix/System.swift [hunk 5]
```diff
@@ -756,7 +759,21 @@ internal enum Posix {
...
+    @inline(never)
+    public static func mkdir(pathname: String, mode: mode_t) throws {
+        _ = try syscall(blocking: false) {
+            sysMkdir(pathname, mode)
+        }
+    }
+
 #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    @inline(never)
+    public static func mkpath(pathname: String, mode: mode_t) throws {
```
Comment (Lukasa) [inline Sources/NIOPosix/System.swift:771]: This function should be called `mkpath_np`.

File: Sources/NIOPosix/NonBlockingFileIO.swift [hunk 6]
```diff
@@ -687,6 +687,66 @@ public struct NonBlockingFileIO: Sendable {
...
+        // Fail fast if not a directory or file exists
+        do {
+            var s = stat()
+            try Posix.stat(pathname: path, outStat: &s)
+            if (S_IFMT & s.st_mode) == S_IFDIR {
+                return
+            }
+            throw IOError(errnoCode: ENOTDIR, reason: "Not a directory")
+        } catch {
+            // ignore error
```
[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/NIOPosix/NonBlockingFileIO.swift:700]: we can catch only `file not found` as it is what we expect and rethrow all other errors, wdyt?

[the remark this answers is not in the document]
Comment (Lukasa) [reply Sources/NIOPosix/NonBlockingFileIO.swift:700]: Seems reasonable.

[the remark this answers is not in the document]
Comment (artemredkin) [reply Sources/NIOPosix/NonBlockingFileIO.swift]: added a catch for `ENOENT` only

File: Sources/NIOPosix/NonBlockingFileIO.swift (hunk 2 above)
Comment (artemredkin) [reply Sources/NIOPosix/NonBlockingFileIO.swift]: added a check in the beginning of the function

File: Sources/NIOPosix/NonBlockingFileIO.swift (hunk 1 above)
Comment (artemredkin) [reply Sources/NIOPosix/NonBlockingFileIO.swift:719]: added comments

File: Sources/NIOPosix/NonBlockingFileIO.swift (hunk 3 above)
Comment (artemredkin) [reply Sources/NIOPosix/NonBlockingFileIO.swift]: fixed

File: Sources/NIOPosix/NonBlockingFileIO.swift (hunk 4 above)
Comment (artemredkin) [reply Sources/NIOPosix/NonBlockingFileIO.swift:736]: added comments

File: Sources/NIOPosix/System.swift (hunk 5 above)
Comment (artemredkin) [reply Sources/NIOPosix/System.swift]: fixed

File: Sources/NIOPosix/NonBlockingFileIO.swift [hunk 7]
```diff
@@ -687,6 +687,82 @@ public struct NonBlockingFileIO: Sendable {
...
+            throw IOError(errnoCode: ENOTDIR, reason: "Not a directory")
+        } catch let error as IOError where error.errnoCode == ENOENT {
+            // if directory does not exist we can proceed with creating it
+        }
+
+        // Slow path, check that all intermediate directories exist recursively
+
+        // Trim any trailing path separators
+        var index = pathView.index(before: pathView.endIndex)
+        let pathSeparator = 47 // 47 is `/`
```
Comment (Lukasa) [inline Sources/NIOPosix/NonBlockingFileIO.swift:712]: Prefer `UInt8(ascii: "/")`

File: Sources/NIOPosix/NonBlockingFileIO.swift [hunk 8]
```diff
@@ -687,6 +687,82 @@ public struct NonBlockingFileIO: Sendable {
...
+        }
+
+        // Find first non-trailing path separator if it exists
+        while index != pathView.startIndex && pathView[index] != pathSeparator {
+            index = pathView.index(before: index)
+        }
+
+        // If non-trailing path separator is found, create parent directory
+        if index > pathView.startIndex {
+            try self.createDirectory0(String(decoding: pathView.prefix(upTo: index), as: UTF8.self), mode: mode)
```
Comment (Lukasa) [inline Sources/NIOPosix/NonBlockingFileIO.swift:724]: Nit: I _think_ you can create a `Substring` from a `String.UTF8View.Subsequence`, and then create a `String`, so you should be able to do:

```
try self.createDirectory0(String(Substring(pathView.prefix(upTo: index))), mode: mode)
```

File: Sources/NIOPosix/NonBlockingFileIO.swift (hunk 7 above)
Comment (artemredkin) [reply Sources/NIOPosix/NonBlockingFileIO.swift]: 👍

File: Sources/NIOPosix/NonBlockingFileIO.swift (hunk 8 above)
Comment (artemredkin) [reply Sources/NIOPosix/NonBlockingFileIO.swift]: ah, fixed

**Review by Lukasa — APPROVED** [review]

**Pull request merged**